### ISSUE TITLE
prettierIgnore key in package.json?

### ISSUE THREAD

**Issue by kylesimmonds** [prettier/prettier#3460]
ESlint currently allows ignore options to be set in `package.json` under the `eslintIgnore` key: https://eslint.org/docs/user-guide/configuring#using-eslintignore-in-packagejson

I noticed there's no mention of a `prettierIgnore` option in the docs (https://prettier.io/docs/en/ignore.html). I also tried adding a similar key for prettier like this, which did nothing:

```
"prettierIgnore": ["build"],
```

At this point I'm pretty sure this isn't possible currently, but I wanted to make sure. Are there plans to add this in or is this intentionally left out?

**Comment by azz** [comment]
It's not currently implemented. We don't currently look at the package.json at all. We use cosmiconfig for configuration which is separate to the ignore file handling.

**Comment by lipis** [comment]
@azz that is not true. You can add the `"prettier"` key in the `package.json` with all the options there.

**Comment by duailibe** [comment]
@lipis Yeah that’s the “we use cosmicconfig” part, but we don’t look the package.json for ignore.. those are two completely separate flows

**Comment by lipis** [comment]
@duailibe I see :)

**Comment by azz** [comment]
What's the incentive for putting this in package.json anyway?

**Comment by kylesimmonds** [comment]
@azz the only reason I can think of is keeping the config format options as close to ESlint as possible because many people use it alongside ESlint and would expect the options to be the same. I could be wrong but not allowing `prettierIgnore` is the only difference from where you can add ESlint settings.

Maybe more people prefer using files to package.json keys (`.eslintrc`, `.eslintignore`, `.prettierrc`, `.prettierignore`) but it seems incomplete that the only thing missing is allowing `prettierIgnore` in package.json

**Comment by azz** [comment]
I don't find "because ESLint does it this way" a particularly compelling reason. I think people like to put all kinds of things in their package.json files that don't really belong, and this seems like one of those things.

**Comment by kylesimmonds** [comment]
@azz you're right that's not a compelling reason, but assuming that it shouldn't belong in package.json is a pretty sweeping dismissal of another dev's potential needs. By that reasoning using package.json shouldn't be an option for `"prettier"` either. And seeing as this is the only config missing in package.json it seems more like an accident than an informed decision.

If you don't want to do it, that's fine, I won't lose sleep over an extra file. But don't attribute it to wanting to keep package.json clean

fwiw, standard also allows exclusions in package.json, though it's a different format: https://standardjs.com/index.html#how-do-i-ignore-files

**Comment by hawkrives** [comment]
I was looking through the documentation for how to add the ignore list to my package.json config a few days ago, actually.

My reasoning is that I don't like adding new files to my top-level folder, and while I can get around that with the `"prettier"` package.json key, I still have to create a `.prettierignore`. 

Idk. Not a _huge_ deal, but a minor annoyance for my psyche.

**Comment by olsonpm** [comment]
yeah I thought the community was moving toward project configurations inside `package.json` because the dot files at the project directory were getting out of hand.  I don't remember reading any functional constraints, more a community wide preference.  But I've seen it plenty other places than just eslint.

**Comment by nickmccurdy** [comment]
Cosmic config supports this by default, so I see no reason to not support this in `package.json` (if it isn't supported already). This doesn't prevent you from using a separate file if you prefer, it's just another option. This is quickly becoming a de facto standard for JSON configuration in JS packages.

**Comment by duailibe** [comment]
Just FYI (this is not a justification) we don't use `cosmiconfig` for ignore, only for the configuration stuff.

**Comment by nickmccurdy** [comment]
Ah I wasn't aware, thanks. Would it be a good idea to use it for ignore or can there be only one config per project?

**Comment by suchipi** [comment]
`cosmiconfig` is built for loading configurations from JSON or yaml, but prettierignore files are a list of file globs and negations, perhaps with comments. You could emulate that behavior with an array or list, but it'd be a little bit awkward and incongruent with eg gitignore files. Also, I think cosmiconfig would look for files named ".prettierignorerc". So it's not really the right tool for this job.

**Comment by nickmccurdy** [comment]
`eslintIgnore` is fine in `package.json`. They use custom logic instead of cosmicconfig to implement it, but it doesn't look very complex. https://github.com/eslint/eslint/blob/master/lib/ignored-paths.js

**Comment by olsonpm** [comment]
> I think cosmiconfig would look for files named ".prettierignorerc". So it's not really the right tool for this job.

cosmiconfig can be configured to look for whatever file names you choose.  You're right by default it includes `*rc` files.

**Comment by cenobitedk** [comment]
> ... project configurations inside `package.json` because the dot files at the project directory were getting out of hand. 

Exactly. I'm working on writing a common ground (boilerplate) for a larger company and the more config I can combine in one place, the better. It will make it more flexible and more transportable.

This issue hasn't changed in a year, what's the status?

**Comment by jamonholmgren** [comment]
Updated link to ESLint implementation. It’s really simple, although they don’t use cosmiconfig for it.

https://github.com/eslint/eslint/blob/master/lib/cli-engine/ignored-paths.js

I’m tempted to make a PR; anyone else interested, though? I have plenty of other open source on my plate.

**Comment by VivekNayyar** [comment]
Interested in making a PR for the same. Do I need to worry about cosmicconfig or can we go how eslint is doing it?

**Comment by ark120202** [comment]
ESLint is going to allow specifying ignore patterns in config (see https://github.com/eslint/eslint/pull/12274), was it considered to do the same for Prettier? That way it would be possible to specify them everywhere it can be configured, including `package.json`

**Comment by kael-shipman** [comment]
I see this is tagged as "discussion needed", but it appears to have gone stale. Is there any way we can get to a definitive yes or no here? This seems to be an expected feature by a significant group of people, including myself, and the path to implementation, while perhaps not as straightforward as hoped, is certainly not treacherous or fraught.

My argument toward "yes" is simply that there is no semantic difference to the _user_ between types of config. From a user perspective, config is config, and any options that appear in the help text (`prettier -h`) should be available in _any_ config file, including `package.json`. While from a program perspective we seem to be drawing a difference between "environment" config (which files do we act on, etc.) and "functional" config (what do we do to those files), that difference is in no way pertinent to the end user, which is for whom we build our products, and revealing that difference is simply confusing.

One more thing to consider: Who does it harm to implement this? Will anyone (users, documenters or project contributors) say "oh my god, this was such a bad idea"? Probably not.... So if there's virtually no cost (after the cost of implementation) and there's significant gain, why not just do it?

**Comment by evilebottnawi** [comment]
I think we should use logic from eslint https://github.com/eslint/eslint/pull/12274, allow to use `ignorePatterns` in configuration

**Comment by kael-shipman** [comment]
@azz? @duailibe ? I know yall are probably tied up in all kinds of other stuff (much appreciated ;)), but if you'd at least consent to development of this harmless feature, we could get to work on a PR. What do you think?

**Comment by azz** [comment]
I wouldn't like to see it in the configuration object like the other API configuration properties, because it can lead to confusion when using the prettier API. `prettier.format` will _not_ respect this configuration.

One possibility is to add it to the configuration under a `cli` object:

```json
{
  "prettier": {
    "singleQuote": true,
    "cli": {
      "ignore": ["build"]
    }
  }
}
```

We'd need to make sure `prettier.getFileInfo` respects such a setting too.

Note that there are [many](https://prettier.io/docs/en/cli.html) other config flags that are _CLI-only_. We'd have to be careful as to which ones are supported in a config file. Some flags act more like directives/commands than switches or configuration.

**Comment by kael-shipman** [comment]
@azz , that's a fair proposal, and I'm probably not the only one who greatly appreciates the compromise.

Regarding the other CLI-only config flags, it looks to me like all prettier config can be broken into 3 basic categories: 1) Formatting options; 2) which files to format (or not); and 3) I/O options (e.g., whether or not to write changes in place, whether to output this or that thing, where config should be read from, etc.). Given that, I would say it makes sense that the first two categories should be available in _all_ areas in which config can be supplied, while the last category is definitely specific to the CLI.

If it helps, perhaps you could group the first two categories together as "functional config", while the last one is interface-specific.

Anyway, I'm perfectly happy with the `cli` flag option. @cinnamonkale , @evilebottnawi , you two also seemed to have some opinions about this topic. Does this solution seem acceptable to you?

If everyone's happy, maybe we can remove the `discussion needed` tag and make it available for someone to implement.

**Comment by lydell** [comment]
See also: #7073 (that proposal includes the ability to write config and ignores in the same file)

**Comment by kael-shipman** [comment]
@lydell , very good initiative (full support!), but I think to bundle this particular issue with the one you've linked would derail what could be some very easy incremental progress here. My vote is to keep the two issues active and separate.

(To be fair, it looks like you're not asking to combine issues, just for this thread to be aware of your initiative, which is fine.)

**Comment by cinnamonkale** [comment]
@kael-shipman @azz the `prettier.cli.ignore` option feels more like a shim to me than allowing a separate `prettierIgnore` item, or even just `prettier.ignore`, but if everyone's cool with that so am I.

As other people have said, including in #7073 , the way ESLint does things isn't necessarily the best way, but as someone who uses these tools everyday I'd rather the behavior match across these tools, even if that behavior could be implemented better. That's of course a preference and not based on anything else

**Comment by kael-shipman** [comment]
Cool, thanks for the input.

My understanding of the situation here is that whatever we may gain by discussing this in depth and finding the absolute best way to make it happen is unlikely to have much of an impact compared to simply having the feature implemented _somehow._ It's a relatively trivial and easy feature, and a number of people would likely benefit from it.

Thus, my thought is that it would be beneficial to go with @azz 's suggestion, since he's a maintainer of the project, and then expect that it will likely become a non-issue when @lydell 's work comes to pass.

@cinnamonkale @azz @lydell @duailibe @evilebottnawi it sounds to me like everyone is more or less willing to go with the `prettier.cli.ignore` idea proposed by @azz . Can we get a vote? If everyone's cool with it, perhaps we can approve this proposal so someone can begin implementing?

**Comment by cmonacaps** [comment]
> Maybe more people prefer using files to package.json keys (.eslintrc, .eslintignore, .prettierrc, .prettierignore) but it seems incomplete that the only thing missing is allowing prettierIgnore in package.json

I think most people just go with what works and don't spend too much time to post about it.

There are a lot of reasons to try for a consolidated build configuration, it's so much more portable in one file, where the config block is along with the dependency that uses it, than the situation where many packages have their own config files.

It makes sense that we have a `.gitignore`, a `Dockerfile`, a `Jenkinsfile` since they are external to the node ecosystem described by `package.json`

Even a `webpack.json` makes sense, it's describing something other than the local dev environment.

Less easy to see why a package in `./node_modules` that everyone uses as a dev dependency, and everyone wants to configure a bit, would have a few important settings requiring an extra file, while it does support most of its configuration being in `package.json`.

**Comment by churchofthought** [comment]
+1

**Comment by churchofthought** [comment]
➕1️⃣

**Comment by nanxiaobei** [comment]
Need this one, `"prettierIgnore"` key in `package.json`

**Comment by benatkin** [comment]
Use case: get prettier to ignore .yaml, because its yaml formatting is unusual.

**Comment by lydell** [comment]
@benatkin That use case is already supported: You can add `*.yaml` to the `.prettierignore` file.

**Comment by christian-draeger** [comment]
any news to add the prettier ignore config to the package.json somehow?
totally agree with @cmonacaps saying
> There are a lot of reasons to try for a consolidated dev configuration, it's so much more portable in one file, where the config block is along with the dependency that uses it, than the situation where many packages have their own config files.

imho the way eslint is doin it by "ignorePatterns" is the most intuitive way plus it would be nice if prettier would do it the same. but in the end i would also go for the cli option thing that @azz recommended. most important is to have the option, regardless regardless of how it is exactly solved

**Comment by fisker** [comment]
Instead of add `prettierIgnore` field, how about add `ignore: []` in `prettier` field, so we can use `cosmiconfig` to find it, I think it's easier to implement.

**Comment by neodon** [comment]
I was trying a new approach today in a project to consolidate all the top-level dotfiles into package.json after seeing examples of others doing it.

It's so much cleaner (in my newly formed opinion) to have package configs in the package config file.

I was surprised that prettier is the only tool I'm using that doesn't support this approach.  I'm even more surprised to see that it's been a controversy for almost 4 years.

Please consider allowing some kind of progress on this issue.  Thank you.

**Comment by DaniBedz** [comment]
Agreed. There's no reason that this shouldn't be supported by now. People like clean codebases. If I can move a config file into package.json, you better believe I will.

**Comment by AntonOfTheWoods** [comment]
This seriously reminds me of https://github.com/PyCQA/flake8/issues/234. Virtually entire community thinks `flake8` config should be able to go in `pyproject.toml`. Maintainer doesn't. Sadness.

**Comment by yangwao** [comment]
I came across this thread to see if it's possible to add to key `.prettierIgnore` to `package.json` as an extra file at top-level is a bit annoying to clear repository. 

```json
  "prettier": {
    "cli": {
      "ignore": [
        ".vscode/"
      ]
    }
  },
```
```json
  "prettier": {
    "cli": {
      "ignorePath": [
        ".vscode/"
      ]
    }
  },
```

EDIT Probably this is way to go
> To exclude files from formatting, create a .prettierignore file in the root of your project. .prettierignore uses gitignore syntax.

**Comment by mataha** [comment]
I've opened a PR [here](12672) - would appreciate some feedback.

**Comment by thilllon** [comment]
check

**Comment by kachkaev** [comment]
I agree with some folks here that we should go for `ignorePatterns` in Prettier config instead of adding more fields to `package.json`. Alignment with ESLint would be nice!

📖 [Docs on `ingorePatterns` in ESLint config](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code#ignorepatterns-in-config-files)

I’ve asked @mataha if they want to continue with https://github.com/prettier/prettier/pull/12672. Either way, it'd be great to see `ignorePatterns` landed!

**Comment by sean-au** [comment]
Adding ignore patterns has also been requested many times here #13088

**Comment by kachkaev** [comment]
@sean-au, yep. I’ve merged https://github.com/prettier/prettier/issues/13088 into this discussion so that we can join forces 🙂

**Comment by valtlai** [comment]
> Alignment with ESLint would be nice!

In the [new config file system](https://eslint.org/docs/latest/user-guide/configuring/configuration-files-new) ([blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/)), the key is called `ignores`, so I think that’s the name we should use.

**Comment by sean-au** [comment]
> > Alignment with ESLint would be nice!
> 
> In the [new config file system](https://eslint.org/docs/latest/user-guide/configuring/configuration-files-new) ([blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/)), the key is called `ignores`, so I think that’s the name we should use.

The issue with ignores (vs ignorePatterns) is you can't copy and paste configuration from eslint or stylelint, which prettier is often used alongside

**Comment by binaryben** [comment]
This is top result for me when searching how to add ignore patterns to package.json. Seems like a reasonable request, why still no action on it? Just the maintainers don't like it so we suck it up?

**Comment by Standard8** [comment]
@binaryben That is unfair and unnecessary. The maintainers might have various reasons that they haven't yet got to this, but we don't know what they are. Do we have to remind you this is open source and free?

No I'm not a maintainer, and yes I want something similar. Rather than complaining, I did something. My first attempt (linked literally just above your comment) had some reasonable concerns raised, and I've a second attempt under way, unfortunately, I too haven't had much time recently.

**Comment by binaryben** [comment]
@Standard8 I meant no offence or complaint, just an honest enquiry about the state of this issue and reasoning for wherever it is where it is. Thanks for pointing out the work you're doing. Most of what I had read was bikeshedding and I'll admit I didn't read past yangwao's comment so I did miss your PR.

**Comment by fisker** [comment]
Hey, I'm a maintainer. Sorry for letting you down.

Personally I want to support this too. But it's not simple to support, for example we need know where is package.json, so we can know what are those patterns related to.

My plan is to add multiple ignore files support first, then treat package.json with ignorePatterns as an `.ignore` file. I didn't post this idea because I'm busy with other works, haven't get time to finish #14332.

**Comment by Standard8** [comment]
> Personally I want to support this too. But it's not simple to support, for example we need know where is package.json, so we can know what are those patterns related to.
> 
> My plan is to add multiple ignore files support first, then treat package.json with ignorePatterns as an `.ignore` file. I didn't post this idea because I'm busy with other works, haven't get time to finish #14332.

Do you mean using `ignorePatterns` within `package.json` or a list of ignore files which is what [ESLint supports](https://eslint.org/docs/latest/use/configure/ignore#using-eslintignore-in-packagejson)?

I'm curious about why we're heading towards `package.json` rather than the prettier configuration file? I would have thought it would be easier to work in the prettier configuration file and more flexible, and [some of the comments](https://github.com/prettier/prettier/issues/3460#issuecomment-587492818) above suggested it might be good to head in that direction as that would also match ESLint.

This was the approach I was heading towards as a follow-up to #14223 although I did come across a few complications with that.

I definitely don't want to derail your efforts though, I'm more checking what you're intended there, and seeing if I should continue my efforts or not (I'm thinking probably not). At the end of the day I'd love any sort of multi-file support, so anything is a bonus.

**Comment by PatrykMilewski** [comment]
I would love to see this feature, in my case we are having tons of config/ignore files in the root of repo, that's why I'm trying to limit them and hide all of that in `package.json`

**Comment by jregistr** [comment]
>What's the incentive for putting this in package.json anyway?

For me, I just mainly wanna reduce the number of `.` files in my repo.

**Comment by wpyoga** [comment]
I agree with @kachkaev , and I would also love to see support for `ignorePatterns` inside prettierrc, just like @ivodolenc mentioned in https://github.com/prettier/prettier/pull/12672#issuecomment-1256947760 .

As an added benefit, with this feature, we can reuse all `.gitignore` files if we use a js-based prettierrc:

```
module.exports = {
  ignorePatterns: require('eslint-gitignore').readGitignoreFiles({ cwd: __dirname }),
  ...
};
```

Note: [eslint-gitignore](https://github.com/mysticatea/eslint-gitignore) actually just reads all `.gitignore` files in the project directory, and generate an array of `.gitignore` entries. It doesn't do anything specific to ESLint. There is also [parse-gitignore](https://github.com/jonschlinkert/parse-gitignore) as an alternative.

On another note, should this feature request be tracked here, or in #4708 ? I feel that they overlap somewhat.

**Comment by davea38** [comment]
Just my two pence, but it would be great to cut down on the various ignore files and be able to maintain one for auto formatting etc. Being as they all use the same format I don't see this as a particular problem. 

Specifying the file path of the ignore file via package.json is a great idea +1 from me!

**Comment by nickmccurdy** [comment]
> Just my two pence, but it would be great to cut down on the various ignore files and be able to maintain a single one (prettier / eslint etc) for auto formatting etc.

You can pass `--ignore-path` to both Prettier and ESLint to set a single ignore file. If you use `.gitignore`, Prettier 3 will use it by default, too.

**Comment by davea38** [comment]
> > Just my two pence, but it would be great to cut down on the various ignore files and be able to maintain a single one (prettier / eslint etc) for auto formatting etc.
> 
> You can pass `--ignore-path` to both Prettier and ESLint to set a single ignore file. If you use `.gitignore`, Prettier 3 will use it by default, too.

Thank you for mentioning this as I didn't know that this was possible.
I'm currently using eslint-plugin-prettier - I don't believe that it allows me to do this sadly :(

**Comment by nickmccurdy** [comment]
> currently using eslint-plugin-prettier - I presume to do this I need to add it as an option in the eslint rule.

No, it should be using the same ignore list as ESLint. That being said, if you still have issues configuring this with `eslint-plugin-prettier`, I'd recommend switching to [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier), which makes it easy to use ESLint and Prettier separately without conflicts.

**Comment by kirkbushell** [comment]
I just stumbled upon this - honestly, I found it really odd that -some- configuration can go in package.json but not ignores. It's intuitive to put it there if that's where you want to put it.

The fact that prettier supports all these different areas only adds to confusion - took me about half an hour to get everything wired up correctly. Some areas suggest prettierrc, others the json file, others in different files together.

**Comment by nickmccurdy** [comment]
I think it's a bit easier to set up Prettier 3 now with `.gitignore`, but I'm still in favor of adding `prettierIgnore` explicitly for consistency.

**Comment by jasongitmail** [comment]
The best location would be within Prettier config, rather than scattered into more files like `package.json`. Removing the need for `.prettierignore`, and eventually deprecating it, would help reduce JS config file pollution/fatigue too. 

Better DX to just have one way and one config file:
> One way to define configs - we didn’t want folks to have multiple ways to do the same thing any longer. There should be one way to define configs for any given project. [ESLint](https://eslint.org/blog/2022/08/new-config-system-part-2/#the-goals-of-flat-config)

**Comment by Standard8** [comment]
Although there's a risk of splitting the discussion, I've just filed #15481 asking for supporting the configuration in the same way as ESLint (i.e. via `ignorePatterns` in the configuration file).

Whilst I respect the desire to have all the configuration in one file, as per the reasons I mention there, `package.json` does not have sufficient capability for anything other than a small project. Hence this would be insufficient for my needs, and so I'm splitting that request out to make it clear what is really needed.

**Comment by devjiwonchoi** [comment]
Hi engineers, opened #16286 for this and added tests. I'd like some reviews before moving on to adding docs. Feel free to take a look, thanks!
